File: backend/src/main/java/com/zlatenov/wedding_backend/service/RsvpService.java
package com.zlatenov.wedding_backend.service;

import com.zlatenov.wedding_backend.dto.FamilyMembersResponse;
import com.zlatenov.wedding_backend.dto.GuestResponse;
import com.zlatenov.wedding_backend.dto.RsvpRequest;
import com.zlatenov.wedding_backend.dto.RsvpResponse;
import jakarta.validation.Valid;

/**
 * @author Angel Zlatenov
 */

public interface RsvpService {

    /**
     * Process an RSVP submission
     *
     * @param request The RSVP request containing guest responses
     * @param username The authenticated username (first and last name)
     * @return Response with success status and information
     */
    RsvpResponse processRsvp(RsvpRequest request, String username);

    /**
     * Get family members for the RSVP form
     *
     * @param name The authenticated username
     * @return Response with primary user and family members information
     */
    FamilyMembersResponse getFamilyMembers(String name);

    /**
     * Get the current RSVP status for a user and their family
     *
     * @param name The authenticated username
     * @return Response with the current RSVP status
     */
    RsvpResponse getRsvpStatus(String name);

    /**
     * Save the primary guest's response only
     * @param guestResponse The primary guest's response
     * @param username The authenticated user's username
     * @return Response with success status
     */
    RsvpResponse savePrimaryGuestResponse(@Valid GuestResponse guestResponse, String username);
}

File: frontend/src/app/rsvp/services/rsvp.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export type ResponseStatus = 'MAYBE' | 'YES' | 'NO';

export interface GuestResponse {
  userId: number;
  status: ResponseStatus;
  dietaryNotes?: string;
  additionalNotes?: string;
}

export interface RsvpRequest {
  primaryGuest: GuestResponse;
  familyMembers?: GuestResponse[];
}

export interface RsvpResponse {
  success: boolean;
  message: string;
  primaryUserId: number;
  confirmedAttendees: number;
}

export interface UserDto {
  id: number;
  firstName: string;
  lastName: string;
  isChild: boolean;
  rsvpStatus: ResponseStatus;
  dietaryNotes?: string;
  additionalNotes?: string;
}

export interface FamilyMembersResponse {
  primaryUser: UserDto;
  familyMembers: UserDto[];
}

@Injectable({
  providedIn: 'root'
})
export class RsvpService {
  private readonly API_URL = '/api/rsvp';

  constructor(private http: HttpClient) {}

  /**
   * Submit RSVP for the current user and their family
   */
  submitRsvp(request: RsvpRequest): Observable<RsvpResponse> {
    return this.http.post<RsvpResponse>(this.API_URL, request);
  }

  /**
   * Get family member details for the authenticated user
   */
  getFamilyMembers(): Observable<FamilyMembersResponse> {
    return this.http.get<FamilyMembersResponse>(`${this.API_URL}/family-members`);
  }

  /**
   * Save the primary guest's response only
   * @param response The primary guest's response
   */
  savePrimaryGuestResponse(response: GuestResponse): Observable<RsvpResponse> {
    return this.http.post<RsvpResponse>(`${this.API_URL}/primary`, response);
  }
}